### PR TITLE
[WFLY-10339]: Broadcast/discovery-group resources have ambiguous requ…

### PR DESCRIPTION
**Pull request by ehsavoie**
…irement specs.

* Adding 6 new resources to have specific boradcast and discovery groups
using socket-bindings or cluster connections
* Updating the schema for these new resources
* Requires connectors for broadcast-group resource

Jira:
https://issues.jboss.org/browse/WFLY-10339
https://issues.jboss.org/browse/WFLY-6607
https://issues.jboss.org/browse/WFLY-11005

### CODE REVIEW

**Comment by ehsavoie** [comment]
@pferraro @jmesnil could you please review this ?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupAdd.java [hunk 1]
```diff
@@ -19,132 +19,73 @@
...
+            OperationStepHandler addHandler;
+            if (operation.hasDefined(JGROUPS_CLUSTER.getName())) {
+                target = target.append(CommonAttributes.JGROUP_DISCOVERY_GROUP, context.getCurrentAddressValue());
+                addHandler = JGroupDiscoveryGroupAdd.LEGACY_INSTANCE;
+            } else {
+                target = target.append(CommonAttributes.SOCKET_DISCOVERY_GROUP, context.getCurrentAddressValue());
+                addHandler = SocketDiscoveryGroupAdd.LEGACY_INSTANCE;
             }
+            op.get(OP_ADDR).set(target.toModelNode());
+            context.addStep(op, addHandler, OperationContext.Stage.MODEL, true);
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupAdd.java]: Why use an intermediate step?  Can't we convert these operations to use their non-legacy versions here?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupAdd.java [hunk 2]
```diff
@@ -19,132 +19,73 @@
...
  */
-public class DiscoveryGroupAdd extends AbstractAddStepHandler {
-
-    public static final DiscoveryGroupAdd INSTANCE = new DiscoveryGroupAdd();
+public class DiscoveryGroupAdd extends ShallowResourceAdd {
 
-    private DiscoveryGroupAdd() {
-        super(DiscoveryGroupDefinition.ATTRIBUTES);
-    }
+    public static final DiscoveryGroupAdd INSTANCE = new DiscoveryGroupAdd(false);
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupAdd.java]: I'm a little confused.  Isn't this handler exclusively for legacy behavior?  When is this non-legacy instance used?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupRemove.java [hunk 3]
```diff
@@ -41,22 +39,34 @@
  */
 public class DiscoveryGroupRemove extends AbstractRemoveStepHandler {
 
-    public static final DiscoveryGroupRemove INSTANCE = new DiscoveryGroupRemove();
+    public static final DiscoveryGroupRemove INSTANCE = new DiscoveryGroupRemove(false);
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupRemove.java]: Same question.  Isn't this inherently a legacy handler?  Why two versions?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupRemove.java [hunk 4]
```diff
@@ -41,22 +39,34 @@
...
+            try {
+                context.readResourceFromRoot(target.append(CommonAttributes.JGROUP_DISCOVERY_GROUP, context.getCurrentAddressValue()));
+                target = target.append(CommonAttributes.JGROUP_DISCOVERY_GROUP, context.getCurrentAddressValue());
+                removeHandler = JGroupDiscoveryGroupRemove.LEGACY_INSTANCE;
+            } catch(Resource.NoSuchResourceException ex) {
+                target = target.append(CommonAttributes.SOCKET_DISCOVERY_GROUP, context.getCurrentAddressValue());
+                removeHandler = SocketDiscoveryGroupRemove.LEGACY_INSTANCE;
+            }
+            op.get(OP_ADDR).set(target.toModelNode());
+            context.addStep(op, removeHandler, OperationContext.Stage.MODEL, true);
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupRemove.java]: Same question.  Why use an intermediate step?  Can we not simply use the non-legacy JGroups/Socket remove handler here?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ *
+ * @author Brian Stansberry (c) 2011 Red Hat Inc.
+ */
+public class JGroupBroadcastGroupRemove extends AbstractRemoveStepHandler {
+    public static final RuntimeCapability<Void> CAPABILITY = RuntimeCapability.Builder.of("org.wildfly.messaging.activemq.broadcast-group", true)
+            .setDynamicNameMapper(DynamicNameMappers.PARENT)
+            .build();
+
+    public static final JGroupBroadcastGroupRemove INSTANCE = new JGroupBroadcastGroupRemove(true);
+    public static final JGroupBroadcastGroupRemove LEGACY_INSTANCE = new JGroupBroadcastGroupRemove(false);
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java]: Same question about legacy vs non-legacy behavior in the same handler.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupAdd.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                // Fabricate a channel resource if it is missing
+                context.addStep(addChannelOperation, AddIfAbsentStepHandler.INSTANCE, OperationContext.Stage.MODEL);
+            }
+        }
+        if(needLegacyCall) {
+            PathAddress target = context.getCurrentAddress().getParent().append(CommonAttributes.BROADCAST_GROUP, context.getCurrentAddressValue());
+            ModelNode op = operation.clone();
+            op.get(OP_ADDR).set(target.toModelNode());
+            context.addStep(op, BroadcastGroupAdd.LEGACY_INSTANCE, OperationContext.Stage.MODEL, true);
+        }
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupAdd.java]: Why is this needed?
If a user issues an add operation for the new resource, we should not need to create any legacy resources.  These would only need to exist if the user relies on commands from the old model version (in which case the add operation would have been issues for the legacy resource).

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupAdd.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                String channel = operation.get(JGROUPS_CLUSTER.getName()).asString();
+                operation.get(JGROUPS_CHANNEL.getName()).set(channel);
+
+                PathAddress channelAddress = context.getCurrentAddress().getParent().getParent().getParent().append(ModelDescriptionConstants.SUBSYSTEM, "jgroups").append("channel", channel);
+                ModelNode addChannelOperation = Util.createAddOperation(channelAddress);
+                addChannelOperation.get("stack").set(operation.get(JGROUPS_CHANNEL_FACTORY.getName()));
+                // Fabricate a channel resource if it is missing
+                context.addStep(addChannelOperation, AddIfAbsentStepHandler.INSTANCE, OperationContext.Stage.MODEL);
+            }
+        }
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupAdd.java]: If we eliminate the deprecated JGROUPS_CHANNEL_FACTORY attribute from this new resource definition, we don't need any special handling here.  All of this logic can remain in the handler for the legacy resource.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import org.jboss.as.controller.capability.DynamicNameMappers;
+import org.jboss.as.controller.capability.RuntimeCapability;
+import org.jboss.dmr.ModelNode;
+
+/**
+ * Removes a broadcast group.
+ *
+ * @author Brian Stansberry (c) 2011 Red Hat Inc.
+ */
+public class JGroupBroadcastGroupRemove extends AbstractRemoveStepHandler {
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java]: Can we just extend ReloadRequiredAddStepHandler?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+    @Override
+    protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+        context.reloadRequired();
+    }
+
+    @Override
+    protected void recoverServices(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+        context.revertReloadRequired();
+    }
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java]: I looks like you can just extend ReloadRequiredAddStepHandler?

**Review by pferraro — CHANGES_REQUESTED** [review]
@ehsavoie In general, the approach here is sound - though legacy handling seems to use an intermediate step that I think can be eliminated.  Additionally, I don't think the new resource definitions need to include any deprecated attributes from the legacy resource they replace.
All of the comments for BroadcastGroup classes will of course apply to their DiscoveryGroup analogs.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+import org.jboss.as.controller.capability.RuntimeCapability;
+import org.jboss.dmr.ModelNode;
+
+/**
+ * Removes a broadcast group.
+ *
+ * @author Brian Stansberry (c) 2011 Red Hat Inc.
+ */
+public class JGroupBroadcastGroupRemove extends AbstractRemoveStepHandler {
+    public static final RuntimeCapability<Void> CAPABILITY = RuntimeCapability.Builder.of("org.wildfly.messaging.activemq.broadcast-group", true)
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java]: yes, this is not used anywhere

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupAdd.java (hunk 6 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupAdd.java]: What if the user mixes old and new commands. We need a 'legacy' resource just to handle the operations

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java (hunk 9 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java]: ok

File: legacy/messaging/src/test/java/org/jboss/as/messaging/test/MigrateTestCase.java [hunk 11]
```diff
@@ -152,19 +152,21 @@ private void testMigrateOperation(boolean addLegacyEntries) throws Exception {
         migrateOp.get(OP_ADDR).add(SUBSYSTEM, SUBSYSTEM_NAME);
 
         ModelNode response = services.executeOperation(migrateOp);
-
+        System.out.println("Response " + response);
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply legacy/messaging/src/test/java/org/jboss/as/messaging/test/MigrateTestCase.java]: yes

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupRemove.java [hunk 12]
```diff
@@ -41,22 +39,34 @@
...
-        ModelNode model = resource.getModel();
-        if (JGROUPS_CLUSTER.resolveModelAttribute(context, model).isDefined() && !JGROUPS_CHANNEL.resolveModelAttribute(context, model).isDefined()) {
-            context.deregisterCapabilityRequirement(ClusteringDefaultRequirement.COMMAND_DISPATCHER_FACTORY.getName(), CAPABILITY.getDynamicName(address));
+    public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+        if (!isLegacyCall) {
+            ModelNode op = operation.clone();
+            PathAddress target = context.getCurrentAddress().getParent();
+            OperationStepHandler removeHandler;
+            try {
+                context.readResourceFromRoot(target.append(CommonAttributes.JGROUP_DISCOVERY_GROUP, context.getCurrentAddressValue()));
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupRemove.java]: yes

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupDefinition.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+            .setAllowExpression(true)
+            .setRestartAllServices()
+            .build();
+
+    @Deprecated
+    public static final SimpleAttributeDefinition JGROUPS_CHANNEL_FACTORY = create(CommonAttributes.JGROUPS_CHANNEL_FACTORY)
+            .setCapabilityReference("org.wildfly.clustering.jgroups.channel-factory")
+            .build();
+
+    public static final SimpleAttributeDefinition JGROUPS_CLUSTER = create(CommonAttributes.JGROUPS_CLUSTER)
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupDefinition.java]: Well I wanted to avoid translating attributes between both handlers

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupDefinition.java [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+import org.jboss.dmr.ModelNode;
+import org.wildfly.clustering.spi.ClusteringDefaultRequirement;
+import org.wildfly.clustering.spi.ClusteringRequirement;
+
+/**
+ * Broadcast group resource definition
+ *
+ * @author <a href="http://jmesnil.net">Jeff Mesnil</a> (c) 2012 Red Hat Inc.
+ */
+public class JGroupBroadcastGroupDefinition extends PersistentResourceDefinition {
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupDefinition.java]: because of the getChildren in ServerResourceDefinition

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupAdd.java (hunk 7 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupAdd.java]: I kept it for compatibility reasons: the legacy operations don't really do something except register a placeholder resource. That's why i kept also the same attribute names to avoid translating them

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupAdd.java (hunk 6 above)
Comment (pferraro) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupAdd.java]: You're right.  We need to handle mixed domains.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java (hunk 5 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java]: legacy means that the handler has been called from a legacy operation command call

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupDefinition.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupDefinition.java]: because i wanted to be able to handle old operations

**Comment by ehsavoie** [comment]
retest this please

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java (hunk 8 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java]: yes

File: legacy/messaging/src/main/java/org/jboss/as/messaging/logging/MessagingLogger.java [hunk 15]
```diff
@@ -819,4 +819,10 @@
 
     @Message(id = 88, value = "Can not migrate attribute failback-delay to resource %s. Artemis detects failback deterministically and it no longer requires to specify a delay for failback to occur.")
     String couldNotMigrateFailbackDelayAttribute(PathAddress address);
+
+    @Message(id = 89, value = "Can not migrate discovery group %s. as no network configuration is properly defined.")
```
Comment (jmesnil) [inline legacy/messaging/src/main/java/org/jboss/as/messaging/logging/MessagingLogger.java]: should be `%s` instead of `%s.`

File: legacy/messaging/src/main/java/org/jboss/as/messaging/logging/MessagingLogger.java [hunk 16]
```diff
@@ -819,4 +819,10 @@
...
+    String couldNotMigrateDiscoveryGroup(PathAddress address);
+
+    @Message(id = 90, value = "Can not migrate broadcast group %s. as no network configuration is properly defined.")
```
Comment (jmesnil) [inline legacy/messaging/src/main/java/org/jboss/as/messaging/logging/MessagingLogger.java]: idem

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ExternalBrokerConfigurationService.java [hunk 17]
```diff
@@ -64,31 +65,31 @@ public void stop(StopContext context) {
     }
 
     public Map<String, TransportConfiguration> getConnectors() {
-        return connectors;
+        return Collections.unmodifiableMap(connectors);
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ExternalBrokerConfigurationService.java]: What's the relation between this change and WFLY-19339?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.ReloadRequiredRemoveStepHandler;
+import org.jboss.dmr.ModelNode;
+
+/**
+ * Removes a broadcast group using JGroups.
+ * @author Emmanuel Hugonnet (c) 2019 Red Hat, Inc.
+ */
+public class JGroupBroadcastGroupRemove extends ReloadRequiredRemoveStepHandler {
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupRemove.java]: As general remark all classes dealing with JGroups should be appended with `JGroup*s*`

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/shallow/ShallowResourceAdd.java [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+import org.jboss.as.controller.registry.Resource;
+import org.jboss.dmr.ModelNode;
+
+/**
+ *
+ * @author Emmanuel Hugonnet (c) 2019 Red Hat, Inc.
+ */
+public abstract class ShallowResourceAdd extends AbstractAddStepHandler {
+
+    protected final boolean isLegacyCall;
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/shallow/ShallowResourceAdd.java]: is this field necessary? this base class does not do anything with it and it seems that it is not related to the "shallow" aspect of the resource.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/shallow/ShallowResourceAdd.java [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+    protected ShallowResourceAdd(boolean isLegacyCall, AttributeDefinition... attributes) {
+        super(attributes);
+        this.isLegacyCall = isLegacyCall;
+    }
+
+    @Override
+    protected void populateModel(final ModelNode operation, final ModelNode model) throws OperationFailedException {
+        //do nothing
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/shallow/ShallowResourceAdd.java]: should this method be empty or should it be mandatory for the concrete subclass to define it?

File: messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties [hunk 21]
```diff
@@ -98,6 +98,8 @@ broadcast-group.start=Starts the broadcast group.
 broadcast-group.started=Whether the broadcast group is started.
 broadcast-group.stop=Stops the broadcast group.
 broadcast-group=A broadcast group is the means by which a server broadcasts connectors over the network.
+jgroup-broadcast-group=A broadcast group is the means by which a server broadcasts connectors over the network using JGroup cluster configuration.
```
Comment (jmesnil) [inline messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties]: JGroups

File: messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties [hunk 22]
```diff
@@ -98,6 +98,8 @@ broadcast-group.start=Starts the broadcast group.
...
+socket-broadcast-group=A broadcast group is the means by which a server broadcasts connectors over the network using socket bindings.
```
Comment (jmesnil) [inline messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties]: UDP socket binding

File: messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties [hunk 23]
```diff
@@ -247,6 +249,8 @@ discovery-group.refresh-timeout=Period the discovery group waits after receiving
 discovery-group.remove=Remove a discovery group.
 discovery-group.socket-binding=The discovery group socket binding.
 discovery-group=Multicast group to listen to receive broadcast from other servers announcing their connectors.
+jgroup-discovery-group=Multicast group to listen to receive broadcast from other servers announcing their connectors using JGroup cluster configuration.
```
Comment (jmesnil) [inline messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties]: idem

File: messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_9_0.xsd [hunk 24]
```diff
@@ -0,0 +1,996 @@
...
+
+            <xs:element name="jgroup-broadcast-group" minOccurs="0" maxOccurs="unbounded">
+                <xs:complexType>
+                    <xs:attribute name="name" type="xs:string" use="required" />
+                    <xs:attribute name="jgroups-stack" type="xs:string" use="optional">
+                        <xs:annotation>
+                            <xs:documentation>Deprecated. Use jgroups-channel instead.</xs:documentation>
+                        </xs:annotation>
+                    </xs:attribute>
+                    <xs:attribute name="jgroups-channel" type="xs:string" use="optional" />
```
Comment (jmesnil) [inline messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_9_0.xsd]: as `jgroups-broadcast-group` is a new resource, we may simplify the name of the attributes and get rid of the `jgroups-` prefix for them (that applies for the management attributes and the XSD attributes)

File: galleon-pack/src/main/resources/feature_groups/messaging-activemq-ha.xml [hunk 25]
```diff
@@ -20,13 +20,13 @@
                 <param name="reconnect-attempts" value="-1"/>
                 <param name="block-on-acknowledge" value="true"/>
             </feature>
-            <feature spec="subsystem.messaging-activemq.server.broadcast-group">
-                <param name="broadcast-group" value="bg-group1"/>
+            <feature spec="subsystem.messaging-activemq.server.jgroup-broadcast-group">
```
Comment (jmesnil) [inline galleon-pack/src/main/resources/feature_groups/messaging-activemq-ha.xml]: jgroups

File: messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_9_0.xsd [hunk 26]
```diff
@@ -0,0 +1,996 @@
...
+    <xs:complexType name="socket-discovery-groupType">
+        <xs:attribute name="name" type="xs:string" use="required" />
+        <xs:attribute name="socket-binding" type="xs:string"  use="required" />
+        <xs:attribute name="refresh-timeout" type="xs:long" />
+        <xs:attribute name="initial-wait-timeout" type="xs:long" />
+    </xs:complexType>
+
+    <xs:complexType name="jgroup-discovery-groupType">
+        <xs:attribute name="name" type="xs:string" use="required" />
+        <xs:attribute name="jgroups-stack" type="xs:string" >
```
Comment (jmesnil) [inline messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_9_0.xsd]: do we need to add a deprecated attribute to a new element?

File: messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_9_0.xsd [hunk 27]
```diff
@@ -0,0 +1,996 @@
...
+                    <xs:complexType>
+                        <xs:attribute name="thread-pool-max-size" type="xs:int" />
+                        <xs:attribute name="scheduled-thread-pool-max-size" type="xs:int" />
+                    </xs:complexType>
+                </xs:element>
+                <xs:element name="http-connector" type="http-connectorType" minOccurs="0" maxOccurs="unbounded" />
+                <xs:element name="remote-connector" type="remote-connectorType" minOccurs="0" maxOccurs="unbounded" />
+                <xs:element name="in-vm-connector" minOccurs="0" maxOccurs="unbounded" type="in-vm-connectorType" />
+                <xs:element name="connector" minOccurs="0" maxOccurs="unbounded" type="connectorType" />
+                <xs:element name="jgroup-discovery-group" type="jgroup-discovery-groupType" minOccurs="0" maxOccurs="unbounded" />
```
Comment (jmesnil) [inline messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_9_0.xsd]: must be named `jgroups-discovery-group`

**Review by jmesnil — CHANGES_REQUESTED** [review]

File: messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_9_0.xsd (hunk 26 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_9_0.xsd]: yes because the old behaviour is still present

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/shallow/ShallowResourceAdd.java (hunk 20 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/shallow/ShallowResourceAdd.java]: no it needs to be empty as the model is not to be filled

File: messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_9_0.xsd (hunk 24 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_9_0.xsd]: Well I wanted to avoid attribute translation. The old methods are still working so they would use the old attribute names. But of course we could have such a Map

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ExternalBrokerConfigurationService.java (hunk 17 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ExternalBrokerConfigurationService.java]: nothing

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/shallow/ShallowResourceAdd.java (hunk 19 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/shallow/ShallowResourceAdd.java]: it is used in the implementations, making it private and adding a protected getter would suit you better ?

Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/shallow/ShallowResourceAdd.java]: ok moving it to the  2 implementations

**Comment by ehsavoie** [comment]
@pferraro @jmesnil all your requested changes have been fixed :) Can you please approve

File: testsuite/integration/vdx/src/test/java/org/wildfly/test/integration/vdx/standalone/MessagingTestCase.java [hunk 28]
```diff
@@ -196,23 +196,24 @@ public void testWrongOrderOfElements() throws Exception {
...
-        assertContains(errorLog, "in-vm-acceptor             shared-store-colocated");
-        assertContains(errorLog, "in-vm-connector            shared-store-master");
-        assertContains(errorLog, "jms-queue                  shared-store-slave");
+        assertContains(errorLog, "divert                     remote-connector");
+        assertContains(errorLog, "grouping-handler           replication-colocated");
+        assertContains(errorLog, "http-acceptor              replication-master");
+        assertContains(errorLog, "http-connector             replication-slave");
+        assertContains(errorLog, "in-vm-acceptor             security-setting");
+        assertContains(errorLog, "in-vm-connector            shared-store-colocated");
+        assertContains(errorLog, "jgroup-broadcast-group     shared-store-master");
```
Comment (pferraro) [inline testsuite/integration/vdx/src/test/java/org/wildfly/test/integration/vdx/standalone/MessagingTestCase.java]: Should be `jgroups-broadcast-group`.

File: testsuite/integration/vdx/src/test/java/org/wildfly/test/integration/vdx/standalone/MessagingTestCase.java [hunk 29]
```diff
@@ -196,23 +196,24 @@ public void testWrongOrderOfElements() throws Exception {
...
+        assertContains(errorLog, "jgroup-discovery-group     shared-store-slave");
```
Comment (pferraro) [inline testsuite/integration/vdx/src/test/java/org/wildfly/test/integration/vdx/standalone/MessagingTestCase.java]: Should be `jgroups-discovery-group`.

**Review by pferraro — CHANGES_REQUESTED** [review]

**Review by jmesnil — APPROVED** [review]

**Comment by jmesnil** [comment]
@pferraro As Emmanuel has addressed all your requested changes, I think this PR can be merged.

**Comment by jmesnil** [comment]
@ehsavoie thanks
